### ISSUE TITLE
BUG: Is offsets.YearBegin working as expected?

### ISSUE THREAD

**Issue by mirkosavasta** [pandas-dev/pandas#52105]
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
pd.to_datetime("2017-09-18") - pd.offsets.YearBegin()
#> Timestamp('2017-01-01 00:00:00') - My expectation was correct as I expected exactly the output of the function

pd.to_datetime("2018-01-01") - pd.offsets.YearBegin()
#> Timestamp('2017-01-01 00:00:00') -  My expectation was wrong as I expected Timestamp('2018-01-01 00:00:00')
```

### Issue Description

Hello,

The documentation for `offsets.YearBegin` states "_DateOffset increments between calendar year begin dates."_
I do not understand if my expectations about the behaviour of `offsets.YearBegin` are correct or not. Please see the code below:

```python
pd.to_datetime("2017-09-18") - pd.offsets.YearBegin()
#> Timestamp('2017-01-01 00:00:00') - My expectation was correct as I expected exactly the output of the function

pd.to_datetime("2018-01-01") - pd.offsets.YearBegin()
#> Timestamp('2017-01-01 00:00:00') -  My expectation was wrong as I expected Timestamp('2018-01-01 00:00:00')
```
Why if I subtract `offsets.YearBegin()` from the first day of the year I get the first day of the previous year? Is this expected? 
If so, I could improve the documentation. If not, I guess this could be a bug.

Thanks in advance.

### Expected Behavior

```python
pd.to_datetime("2018-01-01") - pd.offsets.YearBegin()
#> Timestamp('2018-01-01 00:00:00')
```

### Installed Versions

pandas 1.4.4
Python 3.9.16

**Comment by MarcoGorelli** [comment]
Thanks for the report,

I'd have expected `- pd.offsets.YearBegin(n=0)` to go to the beginning of the current year (just like how `+ pd.offsets.YearBegin(n=0)`) goes to the end of the current year

`+ pd.offsets.YearBegin() - pd.offsets.YearBegin()` seems to work for what you're trying to do:
```python
In [74]: to_datetime(["2017-09-18", "2018-01-01"]) + pd.offsets.YearBegin() - pd.offsets.YearBegin()
Out[74]: DatetimeIndex(['2017-01-01', '2018-01-01'], dtype='datetime64[ns]', freq=None)
```

> Is this expected?
If so, I could improve the documentation. If not, I guess this could be a bug.

**Comment by jbrockmendel** [comment]
instead of addition/subtraction you want rollforward/rollback

**Comment by jbrockmendel** [comment]
Yah we do need rollfoward/rollbackward for arrays, xref #7449.  Also it would help for some of the recent issues with users trying to do `dt64.astype("M8[Y]")` expecting that to round down to the nearest YearBegin.

**Comment by MarcoGorelli** [comment]
Right, the docs for [DateOffset](https://pandas.pydata.org/docs/dev/reference/api/pandas.tseries.offsets.DateOffset.html?highlight=dateoffset#pandas.tseries.offsets.DateOffset) say

> Zero presents a problem. Should it roll forward or back? We arbitrarily have it rollforward:
> date + BDay(0) == BDay.rollforward(date)
> Since 0 is a bit weird, we suggest avoiding its use.

So, we probably shouldn't recommend it in the docs.

From https://github.com/pandas-dev/pandas/issues/7449, it seems the simplest solution is

```
In [28]: to_datetime(["2017-09-18", "2018-01-01"]).to_period('Y').to_timestamp()
Out[28]: DatetimeIndex(['2017-01-01', '2018-01-01'], dtype='datetime64[ns]', freq=None)
```

> If so, I could improve the documentation.

I think it'd be really good to include this example in the `YearBegin` docs (at least, until `rollback` and `rollforward` are available for arrays)

**Comment by MarcoGorelli** [comment]
Re - the docs, it might be good to mirror the `MonthBegin` ones, which mention `rollback` https://pandas.pydata.org/docs/dev/reference/api/pandas.tseries.offsets.MonthBegin.html

**Comment by natmokval** [comment]
I'll add examples to`YearBegin` docs.
